Disable joints when process mode is set to disable.

Ughuuu opened the pull request:
- Fixes https://github.com/godotengine/godot/issues/81756

Also needed if we want breakable joints, as right now there is no good way to disable a joint.

AThousandShips commented:
You will have to add it to the 3D nodes as well before CI can pass, the documentation won't be correct otherwise

Ughuuu commented:
How would I go about disabling the joint? For physics bodies, they get added and removed from the space. For joints, I can call clear_joint and then set it again after that.

Also, I am not getting the notifications called, I am assuming since the individual objects have the _notification callback implemented(eg. PinJoin gets called, but not Joint base class). Should I make a helper and call from each one, or make the Pin call into base?

AThousandShips commented:
To avoid code style errors try using [clang-format locally](https://docs.godotengine.org/en/stable/contributing/development/code_style_guidelines.html#using-clang-format-locally)

Ughuuu committed:
Disable joints when process mode is set for node to disable

fix joint disable cases

fix copy from

update docs

add missing enum

update feedback

remove disable mode clear

upd

fix

update feedback

update feedback part 1

fix build

typo

fix

Update godot_joint_3d.h

Update physics_server_2d_wrap_mt.h

Update joint_3d.h

Update physics_server_3d_extension.h

upd

Update joint_3d.cpp

whitespace diff

lint

Delete sandbox

lint

Ughuuu commented:
Ok, this seems to be working fine. It's ready for another review. Thanks

AThousandShips commented on scene/2d/joint_2d.cpp:
```diff
@@ -62,7 +62,9 @@ void Joint2D::_update_joint(bool p_only_free) {
 	configured = false;
 
-	if (p_only_free || !is_inside_tree()) {
+	bool disabled = is_inside_tree() && !is_enabled() && disable_mode == DISABLE_MODE_REMOVE;
```
```suggestion
	bool disabled = !is_enabled() && disable_mode == DISABLE_MODE_REMOVE;
```
Already covered by the or block, this is only considered if `!is_inside_tree() == false`

AThousandShips commented on scene/3d/joint_3d.cpp:
```diff
@@ -63,7 +63,9 @@ void Joint3D::_update_joint(bool p_only_free) {
 	configured = false;
 
-	if (p_only_free || !is_inside_tree()) {
+	bool disabled = is_inside_tree() && !is_enabled() && disable_mode == DISABLE_MODE_REMOVE;
```
```suggestion
	bool disabled = !is_enabled() && disable_mode == DISABLE_MODE_REMOVE;
```

mihe commented:
I haven't tested this myself, and this might be more of a subjective issue than an objective one, but...

By utilizing `_update_joint(true)` for this (when using `DISABLE_MODE_REMOVE`) you're obviously removing the underlying joint from the server altogether, and thereby also discarding the stored reference frames (relative transforms) for the two bodies. I assume then that once you re-enable the joint the bodies will instead be constrained at whatever reference frames they have at that point in time, rather than the original ones?

I guess it becomes a question of what disabling/enabling a joint really means, or more specifically what the expectation is when enabling a previously disabled joint.

If you did the disabling on a server level instead you would presumably not end up with that "problem", since the underlying joint would be kept intact and instead just omit the impulses that are usually applied. I'm not entirely sure how you go about doing that in Godot Physics, but I would assume it's as simple as just doing an early-out in the [`solve`](https://github.com/godotengine/godot/blob/e3e2528ba7f6e85ac167d687dd6312b35f558591/servers/physics_3d/joints/godot_pin_joint_3d.cpp#L84) method of each joint.

(This also happens to line up better with how Jolt behaves, which allows you to just omit the impulses through its `Constraint::SetEnabled` method, and is what you'd use for breakable joints, so I'm a bit biased here.)

mihe commented:
> I'm not entirely sure how you go about doing that in Godot Physics, but I would assume it's as simple as just doing an early-out in the [`solve`](https://github.com/godotengine/godot/blob/e3e2528ba7f6e85ac167d687dd6312b35f558591/servers/physics_3d/joints/godot_pin_joint_3d.cpp#L84) method of each joint.

I suppose you could also take a similar route to what the collision objects do with regards to being added/removed from the space by adding (on a server level) the ability for joints to add/remove themselves from the body they constrain.

So you'd turn this into a method instead:

https://github.com/godotengine/godot/blob/e3e2528ba7f6e85ac167d687dd6312b35f558591/servers/physics_3d/godot_joint_3d.h#L92-L97

... and I guess you'd need to make this into a method on every joint:

https://github.com/godotengine/godot/blob/e3e2528ba7f6e85ac167d687dd6312b35f558591/servers/physics_3d/joints/godot_pin_joint_3d.cpp#L176-L177

Ughuuu commented:
Not sure tbh what is expected when a joint is enabled after it's disabled. I wouldn't have high hopes for that to do much.

If the objects are by now further away, and you use the initial configurations for it, it will result in physics instability.

If you use the current distance and probably the old configs for the joint, it will work, but the joint will now have a new distance between the nodes. Which will result in more physics stability, but probably not what the user expects.

I think if the joint gets disabled and reenable, it's as if you created the joint from scratch. But it all depends on what your use case is for the joint, I guess, and if you can do both things with what is proposed.

If the joint is always recreated with current bodies transforms, you could just move the bodies to the old positions, and have point 1 solved. But otherwise, if you want to reset the transforms after disable/enable, you have to clear the joint and then set again the old properties.

So yeah, either way we go for, it's gonna be something that we should document I guess, and agree that it's the most obvious choice.

Ughuuu commented:
Let's take specific cases I guess.

If a bridge breaks, and then the bridge goes from:

A)
```
o------o
```

To
B)
```
o.     o
|.     |
|.     |
```

Now, if you reenable the joint, should the bridge try to go back to state A? I would think if that happens, the game developer would want to at least reset the bridge to its initial position.

Or should it, if you enable it and the objects are in weird positions, have a very long joint. It's not ideal, but it's they get what they see. If they want a short joint, they can always teleport the objects close to another and recreate state A.

mihe commented:
> If the objects are by now further away, and you use the initial configurations for it, it will result in physics instability.

This depends on how you set up the joints, to some degree. If you add some springiness to them (or whatever Godot Physics calls it) this presumably becomes less of an issue.

Maybe another disable mode is warranted here? I'm not sure what you'd call it though. If you implemented it like [this](https://github.com/godotengine/godot/pull/81792#issuecomment-1724073605) then calling it `DISABLE_MODE_REMOVE` makes more sense to me, since that lines up better with how `DISABLE_MODE_REMOVE` works for `CollisionObject3D`, where the underlying RID is kept intact. Maybe the currently proposed disable mode could be called `DISABLE_MODE_CLEAR` or `DISABLE_MODE_FREE` instead?

I also think that whatever server-level methods that implements this alternative `DISABLE_MODE_REMOVE` should be used to expose a new `enabled` property on `Joint2D`/`Joint3D`, so that you can disable a joint without needing to completely disable the node processing, and also without disabling its children, but maybe that's better suited for a separate PR.

Ughuuu commented:
I like the idea. Seeing as how this PR is not too big yet, I can do also this, and have the 2 disable modes you talk about.

Will rename then for sake of simplicity to DISABLE_MODE_CLEAR, and take a shot at implementing DISABLE_MODE_REMOVE.

I would also like to make this PR short and implement just this one mode, and rename DISABLE_MODE_REMOVE to DISABLE_MODE_CLEAR, but it feels like it would leave this feature unfinished.

Ughuuu commented:
Ok, updated with a possible implementation of what we talked about. Let me know what you guys think. I am still testing locally different flows.

AThousandShips commented:
You need to generate documentation again

AThousandShips commented:
To make it easier on the history and make squashing easier once you have to do that, make small changes and fixes with `git commit --amend` instead

Ughuuu commented:
Fixed merge conflict after limits and motor change. I will make a small project and a video to show how it works.

AThousandShips commented on servers/physics_2d/godot_joints_2d.cpp:
```diff
@@ -207,7 +211,10 @@ bool GodotPinJoint2D::pre_solve(real_t p_step) {
+	if (!are_constraints_enabled()) {
+		return;
+	}
+	// compute relative velocity
```
```suggestion
	// Compute relative velocity.
```

Ughuuu commented:
Removed the joint_clear as it didn't make much sense. Here is a test project:

[joint-disable.zip](https://github.com/godotengine/godot/files/12879721/joint-disable.zip)

Ughuuu commented:
Any updates on this and if it might get in beta?

AThousandShips commented:
This won't be in 4.2, we're well past feature freeze and new features won't be added, this is scheduled for 4.x

If it isn't marked as scheduled for 4.2 then it's very unlikely to get into 4.2 unless it's a bug fix

Ughuuu commented:
Any updates on this?

rburing commented:
I see the use case but I don't know if this is the best way to implement it, with disabling joints as a feature on the server level, which is unlike the rest of the physics server API. A justification for doing it that way would be that we don't have enough information to re-create a destroyed joint "as it was" (as far as that's possible).

The first class of examples of that type of information is the joint's parameters (in a generalized sense). For example the pin joint's angular limits are relative to the `initial_angle`, which is currently only set when the joint is created and is not configurable, but I think this should be solved by adding an extra `PinJointParam` instead. A second example is the damped spring joint's `rest_length`; if it is not a custom value on the node then it is set to the distance between the anchors *in the physics server*; I think this should be cached on the node instead so that the value can be restored when re-creating the joint in the physics server.

The second class of examples is the internal state saved by the joint, which may be meaningful (and good to keep) or not. To justify the approach taken in this PR, I think we need an example where the keeping of this internal state clearly leads to better behavior than re-creating the joint with all its parameters (including implicit parameters as in the paragraph above). If there is no such example, then I would prefer the approach of just re-creating joints with all their parameters.

I do agree it would be good to also add an `enabled` property to joint nodes (this property and the process mode setter can call the same code), to avoid a joint node's children inheriting the process mode.

Lastly, I am curious how you would implement enabling/disabling joints in the Box2D or Rapier GDExtension with the interface exposed in this PR, when those physics engines don't have this feature as far as I can see. I guess you can fall back to re-creating the joint in those engines, and only actually enable/disable in engines that support it (like Jolt), but IMHO we need an example where this actually leads to better behavior, to be able to approve this PR as is.

Ughuuu commented:
I'm not sure I completely understand, but other physics API do use the disabled flag(or the process_mode), and if you set it to disabled it won't simulate(eg. for rigidbody, and for it you then have DisableMode either remove, make static or keep active). Same for joints, I made it so that if you set process mode to disabled, you have the option to either remove or keep active.

For the two classes of problems, I am assuming these should be handled when enabling/disabling the joints(if something isn't taken into account, it probably should).

I think there is a sort of enable flag, it's a constraint enable flag on the joint.

I was thinking if there is no support for box2d or rapier, to open a issue request and have that feature added.

Let me know how you were thinking this could be done, enabling/disabling of joints(what you are proposing is basically just deleting the joint? So not being able to set it disabled)

rburing commented on doc/classes/Joint3D.xml:
```diff
@@ -31,4 +34,13 @@
 	</members>
+	<constants>
+		<constant name="DISABLE_MODE_REMOVE" value="0" enum="DisableMode">
+			When [member Node.process_mode] is set to [constant Node.PROCESS_MODE_DISABLED], remove from the physics simulation to stop all physics interactions with this [Joint2D].
```
This should say `Joint3D`.

rburing commented on servers/physics_2d/godot_joints_2d.cpp:
```diff
@@ -166,6 +167,9 @@ inline Vector2 custom_cross(const Vector2 &p_vec, real_t p_other) {
 }
 
 bool GodotPinJoint2D::pre_solve(real_t p_step) {
+	if (!are_constraints_enabled()) {
+		return true;
+	}
```
These conditions are unnecessary as the constraints are removed from the bodies, and the physics step gets the constraints from the active bodies. The same goes for `solve`, in this joint and all the other joints, and also in 3D.

rburing commented on servers/physics_2d/godot_joints_2d.h:
```diff
@@ -38,11 +38,39 @@ class GodotJoint2D : public GodotConstraint2D {
+		GodotBody2D *_arr[2] = { nullptr, nullptr };
+	};
+
+	void add_constraints() {
```
I would name this `add_constraint_to_bodies`.

rburing commented on servers/physics_2d/godot_joints_2d.h:
```diff
@@ -38,11 +38,39 @@ class GodotJoint2D : public GodotConstraint2D {
+		}
+	}
+
+	void remove_constraints() {
```
I would name this `remove_constraint_from_bodies`.

rburing commented on servers/physics_2d/godot_joints_2d.h:
```diff
@@ -38,11 +38,39 @@ class GodotJoint2D : public GodotConstraint2D {
 	real_t bias = 0;
 	real_t max_bias = 3.40282e+38;
 	real_t max_force = 3.40282e+38;
+	bool constraints_enabled = true;
```
I would call this just `enabled`.

rburing commented on servers/physics_2d/godot_joints_2d.h:
```diff
@@ -53,36 +81,32 @@ class GodotJoint2D : public GodotConstraint2D {
 	_FORCE_INLINE_ void set_max_bias(real_t p_bias) { max_bias = p_bias; }
 	_FORCE_INLINE_ real_t get_max_bias() const { return max_bias; }
 
+	_FORCE_INLINE_ void set_constraints_enabled(bool p_enabled) {
```
I would name this `set_enabled`.

rburing commented on servers/physics_2d/godot_joints_2d.h:
```diff
@@ -53,36 +81,32 @@ class GodotJoint2D : public GodotConstraint2D {
+			remove_constraints();
+		}
+	}
+	_FORCE_INLINE_ bool are_constraints_enabled() const { return constraints_enabled; }
```
I would name this `is_enabled`.

rburing commented on servers/physics_2d/godot_physics_server_2d.cpp:
```diff
@@ -1042,6 +1042,29 @@ void GodotPhysicsServer2D::joint_clear(RID p_joint) {
 	}
 }
 
+void GodotPhysicsServer2D::joint_set_flag(RID p_joint, JointFlag p_flag, bool p_enabled) {
```
I would name this `joint_set_enabled` and omit the extra flag.

rburing commented on servers/physics_2d/godot_physics_server_2d.cpp:
```diff
@@ -1042,6 +1042,29 @@ void GodotPhysicsServer2D::joint_clear(RID p_joint) {
+	}
+}
+
+bool GodotPhysicsServer2D::joint_get_flag(RID p_joint, JointFlag p_flag) const {
```
I would name this `joint_is_enabled` and omit the extra flag.

rburing commented on servers/physics_3d/godot_joint_3d.h:
```diff
@@ -36,6 +36,32 @@
+
+		GodotBody3D *_arr[2] = { nullptr, nullptr };
+	};
+	bool constraints_enabled = true;
```
Same comments as for 2D.

rburing commented on servers/physics_3d/godot_physics_server_3d.cpp:
```diff
@@ -1220,6 +1220,29 @@ void GodotPhysicsServer3D::joint_clear(RID p_joint) {
 	}
 }
 
+void GodotPhysicsServer3D::joint_set_flag(RID p_joint, JointFlag p_flag, bool p_enabled) {
```
Same comments as for 2D.

rburing reviewed (changes requested):
After a second look this actually mostly looks good, it can just use some simplification as mentioned in the review comments.

To disable a joint it doesn't have to be destroyed, it can just be removed from the simulation (as is done in this PR), as it is done for bodies.

MajorMcDoom commented:
@Ughuuu @mihe  I'm gonna ask this here before starting a new issue.

The current behaviour of physics joints is that if they're removed from the tree (whether on their own or as part of a parent object), when they re-enter the tree, the joint becomes _reconfigured_ using the current state of all objects involved.

IMO, this is very undesirable, because for example if I have a character or some kind of physical mechanism that I'd like to "remove from play", if I do it while the parts are offset from each other, then restoring it the object will completely reconfigure all their joints using a new "neutral", leading to very glitchy behaviour.

The way other engines approach this is for the joints to actually store the offsets for each physics body involved. That way, there is consistency of joint behaviour that survives removal from tree, or disabling. The editor quality-of-life comes from a toggle (default on) for "automatically determining" the offsets based on the current authored positions in the editor. Having this toggle also means you can manually specify offsets in case your objects are not authored in the same configuration as how you'd like them to be joined up at runtime. For example, if you have a bunch of segments you'd like to hook up into a loop, you can author them all in a line.

The current approach in Godot of using the position of the joint to determine the offsets at runtime is very hacky and unreliable. It also makes it impossible to safely readjust offsets at runtime. Imagine you wanted a character's arm to grow - in order to readjust the offsets, you'd have to somehow make sure all relevant parts are aligned in a relative neutral and disable/re-enable the joints in order to trigger their offset calculation routine. But you'd have to also restore all the parts back to their previous transforms.

Anywho, I was going to start a bug report for this, but this PR seems very relevant, so I thought I'd ask here first.

Ughuuu commented:
@MajorMcDoom I do agree with you, I also don't like the way it's currently done, but what you're suggesting is either:
- saving offsets and when re-adding a node using those offsets? (but that could be wrong, what if you add another node. Since you are only given position of a node and you can't change the offset, caching it could be bad. What if you do want to change the length of the joint, you won't be able to, since it's cached.)
- using offset instead of position, but that is a breaking change, and also a change in the API. I would like this more, but am not a maintainer.

Anyway, this PR already is quite big, and has to do more with the what happens when a joint is disabled (right now, if you disable a joint, nothing happens to the joint, it continues to simulate).

MajorMcDoom commented:
> @MajorMcDoom I do agree with you, I also don't like the way it's currently done, but what you're suggesting is either:
> 
>     * saving offsets and when re-adding a node using those offsets? (but that could be wrong, what if you add another node. Since you are only given position of a node and you can't change the offset, caching it could be bad. What if you do want to change the length of the joint, you won't be able to, since it's cached.)
> 
>     * using offset instead of position, but that is a breaking change, and also a change in the API. I would like this more, but am not a maintainer.
> 
> 
> Anyway, this PR already is quite big, and has to do more with the what happens when a joint is disabled (right now, if you disable a joint, nothing happens to the joint, it continues to simulate).

So on the physics server side of things, the offsets already are stored. And in fact, pin joints let you set them explicitly through the server API. The fact that other joints don't let you do so is simply due to not exposing the methods. I think you might be misunderstanding something, because there would be no breaking changes caused by what I'm describing. However I am in agreement that is outside the scope of this PR. I will open up a proposal for this.

Ughuuu commented:
I see, my bad then, was thinking of the API of physics server where it looks like this:
```

	virtual void _joint_make_pin(const RID &p_joint, const Vector2 &p_anchor, const RID &p_body_a, const RID &p_body_b = RID()) override;
	virtual void _joint_make_groove(const RID &p_joint, const Vector2 &p_a_groove1, const Vector2 &p_a_groove2, const Vector2 &p_b_anchor, const RID &p_body_a, const RID &p_body_b) override;
	virtual void _joint_make_damped_spring(const RID &p_joint, const Vector2 &p_anchor_a, const Vector2 &p_anchor_b, const RID &p_body_a, const RID &p_body_b = RID()) override;
```

with the anchor, body_a and body_b.


In code(at least in box2d plugin) I am doing:

```

	Vector2 anchor_A = p_body_a->get_inv_transform().xform(p_pos);
	Vector2 anchor_B = p_body_b ? p_body_b->get_inv_transform().xform(p_pos) : p_pos;

```

Not 100% sure but I thought this was not an offset but a global position. But indeed this is outside of scope, you can create a separate issue for that and it should be solved by a separate PR.

mihe commented:
@MajorMcDoom I've been thinking about something similar to what you're proposing. Exposing the two local-space reference frames (in the case of 3D), as well as actually visualizing them as part of the gizmo, and then having a toggle as to whether you want them automatically calculated or not, would be a more honest representation of how things actually work under the hood, and would provide more flexibility as you point out.

Whether you want to reduce that to something like what Unity does with its "anchor"/"connected anchor" and "axis"/"secondary axis" is a different question I guess.

But yes, this is certainly off-topic for this PR.

Ughuuu commented:
Ok, so @rburing , would these discussions influence the current PR? Have you tested it and the feedback you propose is all needed? I'm assuming anyway won't enter 4.3 so there isn't any rush, especially if there are concerns about this way of doing it.

Ughuuu commented:
Rebased

rburing reviewed (approved):
This looks fine to me now. Thanks for working on this!

Physics server GDExtensions should `override` the new virtual methods for 4.4.

If you use an older physics server GDExtension which does not override these virtual methods then if you disable joints with the default disable mode in 4.4 you will get error messages about those virtual methods not being overridden, and the behavior will be the same as before this PR.

Ughuuu commented:
Forgot to post a video of it working, here it is:

https://github.com/user-attachments/assets/02968218-79ca-4115-977d-1462dd837fe9

Code is:

```gdscript
extends PinJoint2D

func _on_timer_timeout() -> void:
	if process_mode == Node.PROCESS_MODE_ALWAYS:
		process_mode = Node.PROCESS_MODE_DISABLED
	else:
		process_mode = Node.PROCESS_MODE_ALWAYS
```

With a timer that times out every 1s.

Ughuuu commented:
For futher references, last word I got is that this PR still needs work, but I decided not to dedicate any of my time on maintianing it, but instead just adding it myself on a fork of Godot and using it there.

lucasmaffazioli commented:
Thanks for the proposal Ughuuu, was looking forward this PR. Is the fork Blazium, Redot, or private?

Ughuuu commented:
Blazium, will add it to it these following days and merge it there. Thanks for asking.

hpvb commented:
> For further references, last word I got is that this PR still needs work, but I decided not to dedicate any of my time on maintianing it, but instead just adding it myself on a fork of Godot and using it there. Sorry for wasting time.

That's not entirely true, you came to ask what was up, I said that the status of the pr said that it needed review, but that I suspected that it was just an oversight.

I suggested you post in this PR so someone would look at the status, but you closed the PR instead, and wrote a much longer message.

I'm a little confused about this turn of events to be entirely honest.

Ughuuu commented:
Thanks for the message @hpvb, but the care about this PR specifically is not here, not sure it is, but it is just neglected by Godot team. As I said above, my PR has last updates from Nov (and it is already a very old PR, 2023). It isn't merged yet, I have no clue as for what happens. New year up, new dev version, my PR is not there. In my eyes, I have 0 visibility as to what happens, I just have to wait and pray that my free code I contribute for free gets maybe accepted into the engine.
It's not two ways at all, I contribute all the code, I put all the effort, and I am then ignored.
Secondly, about your response message, saying my message is not entirely true. Please know that all of this is a communication issue from Godot team, as I said, I have no clue what happens to my PR from Nov. The normal way things go with this PR so far is. It's reviewed, then someone else merges something, I need to rebase and fix conflicts. Always. It never gets merged, I never get an estimate, no nothing. Now you wanna come along saying what I said is not true, but is it then false? Did I ever get an estimation on the review, a what is going on? I even put an update gif showing the feature working, expecting some kind of update.
Jolt for example, which is also a physics feature, got a preferential treatment. This PR is already reviewed and approved, but it is not merged. Looking objectively at the message Godot sends to me is:
- PR reviewed
- PR approved
- (now what?)
There is 0 communication on this. I wait and see other PR's getting merged, why others are getting merged and not this?

And you say what I said is not entirely true. Then tell me the truth. What is going on, why can't I get any kind of response to my PR. Again I contribute for free my time to this, but my time is ignored.

And for my last point on what I wrote on the chat (which btw, having the rocket chat, if you ask me, just complicates things even more). I asked there in the past for reviews on this, but last time I was ignored also for some time, and I asked, I got messages from @Faless saying I should stop it with the name calling. Sure maybe I was a bit heated about this being ignored, but if you ask me, it's because I cared. So at that point, as I said on the chat, I was thinking, oh, if i write there again, I might get banned, who knows. Also, how do you think I feel, keep asking about this PR getting reviews over and over. When I asked sometime there also, again some other maintainer said I should be careful of contributors (@rburing) time, as they are doing this out of the greatfulnes of their heart. So again, sorry for taking time from everyones busy schedule.

But please do not say what I said is not true. If I am coming to a chat, asking for status after months, and you say the status is needs review, and it is not, then better not say anything. Also, instead of admitting to a mistake, you chose to say I am wrong/lying and saying things that are not true.

I have now checked the conversion and I see @rburing answered after you saying that what you said is not true. So what I said not being true (a lie) is not correct. What I said in the moment I saw what you wrote is correct.

I believe what happens here, including the likes the Godot team choses to stay behind and not write anything, perfectly sums up the experience with Godot. I take my time to write a fully working PR, adding correct message, everything.
And then the Godot team takes 1 minute to look at my stuff, and say something incomplete that causes confusion. Not the same amount of dedication/effort put in from both sides. And it shows. Had I not commented on the chat, this would have not been merged ever, and it will not be merged ever now since I closed it. But form Nov when was last update, no new word, nothing. I believe easily feb, march, etc. Could have come and no new update on this, easily. And this is not the only PR that suffers from this, this is just something that affects me because I thought if I write the code and wait enough, it might get merged sometime. But aparently not.

hpvb commented:
@Ughuuu the PR got approved, but the person who approved it forgot to change the status of the PR in the merge queue. This is why it wasn't merged, someone forgot something.

The thing that I was replying to is simply that I, personally, told you this on rocket.chat but instead of understanding that someone made a mistake in November that we can fix you got mad.

You deleted the thread on rocket.chat in which you asked the question, and got the reply, and now you're here in public accusing us of not caring about your PR.

I'm very sorry that a mistake happened here, genuinely that sucks you put a lot of work in it.

But I don't understand your reaction here, they happened after I personally told you that it was a mistake.

Ughuuu commented:
**Addressing previous comments**:

About the thread that got deleted, sorry about that, it got deleted the moment I deleted my rocketchat account(got frustrated and deleted it). I did that because of the problem that arose here, and the reasons I listed above.

Thanks for the apology, I understand the fact that it was a mistake, I just felt the whole thing was very much not caring about me one bit, thats why my reaction. Thats it. I got a little upset and ur response just made me more angry towards the whole thing.

I'm sorry people that receive what I said with negativity or dislike, it's just my thoughts at the moment, cannot change what I think.

**Addressing PR approval state**

Even thought it is this close to getting finished, I am sure once I start the process there will be other stuff needed, and so on. 

My whole problem with this whole thing is, and always has been, lack of communication, of processes, of expectations. It is specifically on the physics side of godot. There is no meetings about it, no transparency, no direct response, no nothing. On the godot-cpp side, for eg. the things just work where the excelent @dsnopek is in charge. On physics side, there is no maintainer, no meeting, no direct answer, no nothing, so things are just taking longer. On the rocket chat almost always for important stuff the answer was (well, but we need to wait for a maintainer for this, for that, etc). But I would rather not contribute my time to this unless I see a full time physics maintainer on godot side that is taking things serious. As I said above, I have taken this seriously, and not as a joke, and I believe not the same amount of seriousness is there from Godot side.

If anyone wants to take this PR and put it into godot, fine. But it won't be me. As I said I really want to use this feature, so waiting another few months is not something I want (I was really really hoping to get it in the latest dev build, but no, sadly, so many dev builds where I just hoped maybe this time, it might get in).

If/When godot team does want to take this seriously, and not block everything for no reason, I might be re-interested to contribute. But in meantime, it is unrealistic to believe that I will just be like, oh, i'll jsut instantly forget all that happened and continue on. No, some things need to change. I know I am just 1 person and I am not important in the grand scheme of things, but this whole experience has made me less willing to contribute with godot. It's sad but in my eyes, its just this great community and code, but it's all gatekept by some people where everything they do they do, but everything someone else does is met with ignorance.

akien-mga commented:
I understand the pent up frustration after working on a PR for over a year and seeing that it's still not merged, even after it has been approved two months ago.

There was however nothing personal in it taking a long time to review and merge.
Your frustration is understandable, but the anger and negativity addressed towards contributors is unconstructive. I would suggest to take a step back if you're upset.

Now about the lifecycle of this PR specifically:

Physics is an area where we have very few active contributors, and fewer maintainers with the knowledge to evaluate all aspects of proposed physics changes. That's why this PR took so long to reach a first approval.

An approval makes a PR a candidate for merging, and it was tracked as such. The reason it hasn't been merged yet was that we were focusing on the integration of Jolt, following our project priorities. I evaluated this as a change that may impact the Jolt integration too, so I wanted @mihe to also give his approval as he had been involved in discussing the solution. He hadn't gotten to it yet due to working on the Jolt integration and bugfixing.

It was my mistake not to communicate clearly in the PR that I asked for that review, and that it was (a priori) the last requirement before merging the PR.

We're aware that both the PR review process, and our communication on the status of PRs, are lacking. We have a much bigger volume of incoming contributions than contributors available to review them, so there's naturally a bottleneck here - yet we merge around 20 PRs per day, but that's less than what gets opened so some PRs do fall through the cracks and take a longer time to review and merge.

There are no easy solutions there, but we're consistently working on improving both our workflows and our communication on issues, proposals, and PRs.
We understand that the current state of our workflows is not yet "good enough" (neither for contributors who wait too long nor maintainers who are spread too thin), and it's totally fair for some to decide that their time is best spent otherwise.

We're thankful for the work done so far, and understand if you don't want to keep championing this PR. We do think this change addresses a valid issue, so we hope that another physics contributor can take it over (and credit you as co-author).

Ughuuu commented:
Thank you for the nicely written response. I am grateful that the time was taken for this and that this problem was at least addressed. Thanks for shedding some light over the process and what happened to this PR and why it was in what state it was.

I am looking forward to what you wrote about this whole thing improving from the not "good enough" state to hopefully a state that is acceptable.

For me personally,I will try seeing this as a temporary state in which Godot is, but one in which it is nevertheless, but also be understanding they it is something that it actively tries to get out of. I am referring to the physics team lack of people and PR's adding up, which as you hinted is not maintainable long term probably. In any case, I am hopeful that the day will come when such a change/improvement will happen, and am rooting for Godot to get into that state.

In any case I am the only one with this problem, then sorry for causing a lot of distress. I really wanted my concerns voiced, and while some of them may not have been constructive towards the Godot community, the answers I got today were helpful towards me, so thanks for the carefully crafted response.

As a final metaphor I leave the following comparison I see in my head as to how I see this situation (I know some cultures see this things differently, thats why I exaggerated times here) and how I felt about it:
- I invite a friend to a restaurant at 2PM. They agree. I arrive at the restaurant, then I order food. I stay there and wait 2 hours, but they do not arrive. I call them, they do not pick up. After 3 hours they show up (at 5PM), the meal is cold and the restaurant is angry towards me for waiting so long on the table.
